packages/relay/src/lib/clients/cache/impl/customLRUCache.ts

```diff
@@ -0,0 +1,172 @@
+      return false;
+    }
+    return super.delete(key);
+  }
+
+  /**
+   * Deletes a key from the cache without checking if it is protected.
+   * @param {K} key - The key to delete.
+   * @returns {boolean} - True if the key was deleted, false otherwise.
+   * @template K - The key type.
+   */
+  deleteUnsafe(key: K): boolean {
+    return super.delete(key);
+  }
+
+  /**
+   * Loads the pre-configured spending plans from the spending plans configuration file.
+   * @returns {SpendingPlanConfig[]} - The pre-configured spending plans.
+   * @private
+   */
+  private loadSpendingPlansConfig(): SpendingPlanConfig[] {
+    const configPath = findConfig(this.SPENDING_PLANS_CONFIG_FILE);
+    if (!configPath || !fs.existsSync(configPath)) {
+      this.logger.trace(`Configuration file not found at path "${configPath ?? this.SPENDING_PLANS_CONFIG_FILE}"`);
+      return [];
+    }
+    try {
+      const rawData = fs.readFileSync(configPath, 'utf-8');
+      return JSON.parse(rawData) as SpendingPlanConfig[];
+    } catch (error: any) {
+      this.logger.error(`Failed to parse JSON from ${configPath}: ${error.message}`);
+      return [];
+    }
+  }
+
+  /**
+   * Determines if a key is protected. A key is protected if it is associated with a pre-configured spending plan.
+   * @param {string} key - The key to check.
+   * @returns {boolean} - True if the key is protected, false otherwise.
+   * @private
+   */
+  private isKeyProtected(key: string): boolean {
+    if (this.isHbarSpendingPlanRepositoryKey(key)) {
+      return this.spendingPlansConfig.some((plan) => {
+        return (
+          this.isPreconfiguredPlanKey(key, plan) ||
+          this.isPreconfiguredEthAddressKey(key, plan) ||
+          this.isPreconfiguredIpAddressKey(key, plan)
+        );
+      });
+    }
+    return false;
+  }
+
+  /**
+   * Determines if a key is associated with any of the spending plan repositories.
+   * @param {string} key - The key to check.
+   * @returns {boolean} - True if the key is associated with a spending plan repository, false otherwise.
+   * @private
+   */
+  private isHbarSpendingPlanRepositoryKey(key: string): boolean {
+    if (!key) {
+      return false;
+    }
+    return (
+      key.startsWith(HbarSpendingPlanRepository.collectionKey) ||
+      key.startsWith(EthAddressHbarSpendingPlanRepository.collectionKey) ||
+      key.startsWith(IPAddressHbarSpendingPlanRepository.collectionKey)
+    );
+  }
+
+  /**
+   * Determines if a key is associated with a pre-configured spending plan.
+   * @param {string} key - The key to check.
+   * @param {SpendingPlanConfig} plan - The spending plan to check against.
+   * @returns {boolean} - True if the key is associated with the spending plan, false otherwise.
+   * @private
+   */
+  private isPreconfiguredPlanKey(key: string, plan: SpendingPlanConfig): boolean {
+    return key.includes(`${HbarSpendingPlanRepository.collectionKey}:${plan.id}`);
```

Comment:
Nit: This should work since we're working with fixed length keys.  In general I would try to get an exact match, in the case that a key contains the substring of another key.
```
  const expectedKey = `${HbarSpendingPlanRepository.collectionKey}:${plan.id}`;
  return key === expectedKey;
  ```

Reply:
Changed to a more general approach where we initialize a set of protected keys in the constructor and then in `delete` we check for the existence of the key in this set of protected keys:
```typescript
  delete(key: K): boolean {
    if (this.protectedKeys.has(key)) {
      this.logger.trace(`Deletion of key ${key} is ignored as it is protected.`);
      return false;
    }
    return super.delete(key);
  }
```